README.md
# Foxlab Pi - Installer for GitLab CE on a Raspberry Pi 4 (4GB)
This shell script will automate the process of installing GitLab CE 12.6.2 for the Raspberry Pi as well as configuring the install so that it will run properly. The [FoxLab Pi Wiki](https://github.com/k0d3x8its/FoxLabPi/wiki) has manual instructions for installing GitLab CE 12.6.2 as well as a [video tutorial](https://www.youtube.com/watch?v=VVp0buV-wVM) provided by Mikrocontoller Projekte. 
<br> 

<img align="center" src="https://github.com/k0d3x8its/FoxLabPi/blob/master/images/FoxLab-Pi_Menu.png"> 
<br>

A big thanks goes out to <a href="https://github.com/MikrocontollerProjekte">Mikrocontoller Projekte</a> for creating the <a href="https://github.com/MikrocontollerProjekte/GitLabRaspberryPi">GitLabRaspberryPi</a>. 


CHANGELOG.md
# Changelog

## v1.0.0 (2020-05-28)
- **ADDED:** color to the menu
- **REMOVED:** GitLabRaspberryPi.sh file
- **REMOVED:** image part1.png
- **REMOVED:** image part2.png

## v0.9.9 (2020-05-27)
- **ADDED:** wiki for help with installation manually as well as for using FoxLab Pi

## v0.9.8 (2020-05-23)
- **CHANGED:** format of git_data_dirs
- **ADDED:** download and install GitLab menu option

## v0.9.7 (2020-05-22)
- **ADDED:** swapsize menu option

## v0.9.6 (2020-05-20)
- **ADDED:** USB storage menu option
- **REMOVED:** apt from install process in GitLabRaspberryPi.sh

## v0.9.5 (2020-05-20)
- **CHANGED:** position of password change from *3* to *2*
- **ADDED:** password change menu option

## v0.9.4 (2020-05-20)
- **CHANGED:** `blkid` command to `blkid /dev/sda1`
- **ADDED:** hostname menu option
- **ADDED:** sudo when setting hostname

## v0.9.3 (2020-05-13)
- **CHANGED:** shell script from FoxLabPi.sh to foxlab-pi.sh
- **ADDED:** menu framework in FoxLabPi.sh

## v0.9.2a (2020-05-09)
- **ADDED:** fox ascii art to FoxLabPi.sh

## v0.9.1a (2020-05-08)
- **CHANGED:** install to `dpkg -i` in GitLabRaspberyPi.sh
- **ADDED:** ascii art .txt file for fox 
- **ADDED:** FoxLabPi.sh


FoxLabPi.sh
#!/bin/bash

#This shell script is meant to automate MikrocontollerProjekte's 
#instructions for installing GitLab CE on Raspberry Pi 4 (4GB)

#colors
RED='\e[38;5;196m'        #red
ORANGE='\e[38;5;208m'     #orange
YELLOW='\e[38;5;93m'      #yellow
WHITE='\e[38;5;15m'       #white

#Background colors
REDB='\e[41m'             #red

#text manipulation
NT='\e[0m'                #normal text
BOLD='\e[1m'              #bold text
UL='\e[4m'                #underlined text
BLINK='\e[5m'             #blink text


#################
### functions ###
#################

#this will return the user to the menu
pause()
{
    read -p "Press [Enter] to continue down the foxhole..."
}

#this choice will allow the script to update system
updateSystem()
{
    sudo apt update && sudo apt upgrade

    pause 
}

#this choice will allow the script to change hostname and password
hostnameChange()
{
    echo "Hostname Change"

    pause
}

#this choice will mount usb storage
mountStorage()
{
    echo "Mount Storage"

    pause
}

#this choice will increase STACK size
stackSize()
{
    echo STACK Size

    pause
}

#this choice will download GitLab and install
gitLabInstall()
{
    echo "GitLab Download and Install"

    pause
}

#this choice will check if gitlab is running
gitLabCheck()
{
    echo "GitLab Check"

    pause
}


#this will display the choices in a menu format
foxLabMenu()
{
    clear                     

    #FoxLabPi ASCII Art
    echo -e ${ORANGE}"            *.                  *."
    echo -e ${ORANGE}"           ***                 ***"
    echo -e ${ORANGE}"          *****               *****"
    echo -e ${ORANGE}"         .******             *******"
    echo -e ${ORANGE}"         ********            ********"
    echo -e ${ORANGE}"        ,,,,,,,,,***********,,,,,,,,,"
    echo -e ${ORANGE}"       ,,,,,,,,,,,*********,,,,,,,,,,,"
    echo -e ${ORANGE}"       .,,,,,,,,,,,*******,,,,,,,,,,,,"
    echo -e ${ORANGE}"           ,,,,,,,,,*****,,,,,,,,,."
    echo -e ${ORANGE}"              ,,,,,,,****,,,,,,"
    echo -e ${ORANGE}"                 .,,,***,,,,"
    echo -e ${ORANGE}"                     ,*,."

    echo -e ${ORANGE}"    ___           __       _         ${RED}___ _"
    echo -e ${ORANGE}"   / __\____  __ / /  __ _| |__     ${RED}/ _ (_)"
    echo -e ${ORANGE}"  / _\/ _ \ \/ // /  / _  |  _ \   ${RED}/ /_)/ |"
    echo -e ${ORANGE}" / / | (_) >  </ /__| (_| | |_) | ${RED}/ ___/| |"
    echo -e ${ORANGE}" \/   \___/_/\_\____/\__,_|_.__/  ${RED}\/    |_|"${NT}

    echo "####################################################"
    echo "  ----+++++++---- FoxLab Pi Menu ----+++++++----    "
    echo "####################################################"
    echo "1) Update System - - - - - - - - - - -(optional)"
    echo "2) Change Hostname & Password - - - - (Optional)"
    echo "3) Mount USB Storage - - - - - - - - -(Recommended)"
    echo "4) Increase STACK size - - - - - - - -(Recommended)"
    echo "5) Download & install GitLab - - - - -(Necessary)"
    echo "6) Check that GitLab is running - - - (Optional)"
    echo "7) Exit"
}

#this will read the users input
readOptions()
{
    local userChoice
    read -p "Which foxhole do you want to tumble down..." userChoice
    case $userChoice in
        1) updateSystem ;;
        2) hostnameChange ;;
        3) mountStorage ;;
        4) stackSize ;;
        5) gitLabInstall ;;
        6) gitLabCheck ;;
        7) exit 0 ;;
        *) echo -e "${BLINK}${REDB}${WHITE}WARNING${NT}${ORANGE}...You have traveled down the wrong foxhole"${NT} && sleep 6
    esac
}

while true
do
    foxLabMenu
    readOptions
done


GitLabRaspberryPi.sh
# 
#   Installation of GitLab CE on a Raspberry Pi 4 (4GB)
#   - Install GitLab on a Raspberry Pi 4
#   - Configure the GitLab server
#
#   MikrocontrollerProjekte 2020
#   https://www.youtube.com/watch?v=VVp0buV-wVM
#
#   Visit my GitHub page to download this file:	
#   https://github.com/MikrocontollerProjekte/GitLabRaspberryPi/GitLabRaspberryPi.sh
#


################## update your system (OPTIONAL STEP) ################## 

# update your systems package list
sudo apt update

# upgrade all your installed packages to their latest versions
sudo apt upgrade



################## change hostname and password (OPTIONAL STEP) ################## 

# show current hostname
hostname

# open Raspberry Pi configuration tool, change hostname and default user password, enable SSH
sudo raspi-config

# restart Raspberry Pi
sudo reboot

# check hostname
hostname



##################  mount USB Storage (RECOMMENDED STEP) ##################

# creates a directory for the USB Storage
sudo mkdir /media/foxlab

# change owner
sudo chown pi /media/foxlab

# show devices
sudo fdisk -l

# format the USB Storage in EXT4 
sudo mkfs.ext4 /dev/sda1 

# mount the USB Storage
sudo mount /dev/sda1 /media/foxlab

# get the USB Storage UUID (Universally Unique Identifier)
sudo blkid /dev/sda1

# edit fstab
sudo nano /etc/fstab

	# add automount USB Storage on startup 
	UUID=a322538a-2390-4d4f-b13a-7f057b4f2117  /media/foxlab  ext4  defaults  0  0



################## increase STACK size (RECOMMENDED STEP) ##################

# show RAM and STACK usage
free -m

# increase STACK size: CONF_SWAPSIZE=2048
sudo nano /etc/dphys-swapfile

# restart STACK 
sudo /etc/init.d/dphys-swapfile restart

# check available STACK
free -m



################## install GitLab (NECESSARY STEP) ##################

# download the latest package of GitLab CE from https://packages.gitlab.com/gitlab/raspberry-pi2
curl -Lo gitlab-ce_12.6.2-ce.0_armhf.deb https://packages.gitlab.com/gitlab/raspberry-pi2/packages/raspbian/stretch/gitlab-ce_12.6.2-ce.0_armhf.deb/download.deb

# install the downloaded package
sudo dpkg -i gitlab-ce_12.6.2-ce.0_armhf.deb

# modify gitlab.rb configuration
sudo nano /etc/gitlab/gitlab.rb

	# change external_url to the IP of your Raspberry Pi
	external_url 'http://192.168.178.35'

	# reduce the number of running workers to the minimum in order to reduce memory usage (line 750)
	unicorn['worker_processes'] = 2

	# tune the amount of concurrency in your Sidekiq process (line 814)
	sidekiq['concurrency'] = 9

	# storing GitLab data in an alternative directory (line 438)
	git_data_dirs({ 
		"default" => { 
		  "path" => "/media/foxlab/git-data" } })

# reconfigure GitLab
sudo gitlab-ctl reconfigure

# visite http://gitlab/ with your Browser
http://gitlab/



################## check if GitLab is running after reboot (OPTIONAL STEP) ##################

# get GitLab service status
sudo gitlab-ctl status

# stop all GitLab components
sudo gitlab-ctl stop

# restart Raspberry Pi
sudo reboot






#-------------------------------------------------------------------------------------------------------------------


#
#   Usage of GitLab
#   - Generate a SSH key pair for authentication
#   - Create and push a project to your GitLab server
#   - Clone a project from your GitLab server
#   - Issues and merge requests example
#
#   MikrocontrollerProjekte 2020
#   https://www.youtube.com/c/MikrocontrollerProjekte
#
#   Visit my GitHub page to download this file:	
#   https://github.com/MikrocontollerProjekte/GitLabRaspberryPi
#
#   system requirements:
#   - install Git (free and open source distributed version control system):  https://git-scm.com/
#


################## SSH Key (RECOMMENDED STEP) ##################

# create an SSH Key for your server (press enter to don't use a keyphrase)
ssh-keygen -t rsa -b 4096 -C "JohnDoe@gmail.com"

#  ensure that ssh-agent is enabled by running
eval $(ssh-agent -s)

# add your private key to the SSH registry 
ssh-add ~/.ssh/id_rsa

# go to SSH directory
cd ~/.ssh/

# create the config file
touch config



################## push a (STM32CubeIDE) project to your GitLab server ##################

# create a new project in GitLab (new project button or plus icon in the navigation bar) http://gitlab/

# copy the SSH repository URL from the GitLab project (Clone button --> Clone with SSH --> Copy URL to clipboard)

# initialize the local STM32CubeIDE project folder as an empty Git repository
git init

# add the copied SSH repository URL as remote repository where your local repository will be pushed to
git remote add origin git@192.168.178.35:root/stm32f7_motorctrl.git

# verify the new remote SSH repository URL
git remote -v

# add all files to your local repository staging area
git add .

# commit all files that you have staged in your local repository
git commit -m "first commit - initial project"

# push the changes in your local repository to GitLab
git push origin master



################## clone a (STM32CubeIDE) project from your GitLab server ##################

# copy the SSH repository URL from the GitLab project (Clone button --> Clone with SSH --> Copy URL to clipboard)

# clone the project repository into your local STM32CubeIDE workspace
git clone git@gitlab:stm32f7_motorctrl.git

# import the project into STM32CubeIDE workspace



################## push a development branch to your GitLab server and merge it into master branch in GitLab ##################

# create the development branch "dev_startup_delay"
git branch dev_startup_delay

# switch to the development branch 
git checkout dev_startup_delay

# show the working tree status
git status

#  add the changed main.c file to your local repository staging area
git add Core/Scr/main.c

# check the working tree status
git status

# commit changes that you have staged in your local repository
git commit -m "add a startup delay"

# push the development branch in your local repository to GitLab
git push origin dev_startup_delay

# merge the development branch into master branch in GitLab

# switch back to master branch 
git checkout master

# pull new master to local repository
git pull origin master





foxlab-pi.sh
#!/bin/bash

#This shell script is meant to automate MikrocontollerProjekte's 
#instructions for installing GitLab CE on Raspberry Pi 4 (4GB)

#colors
RED='\e[38;5;196m'        #red
ORANGE='\e[38;5;208m'     #orange
YELLOW='\e[38;5;11m'      #yellow
WHITE='\e[38;5;15m'       #white
PURPLE='\e[35m'           #purple
LITEPURPLE='\e[95m'       #light purple
LITEGREY='\e[37m'         #light grey

#Background colors
REDB='\e[41m'             #red

#text manipulation
NT='\e[0m'                #normal text
BOLD='\e[1m'              #bold text
UL='\e[4m'                #underlined text
BLINK='\e[5m'             #blink text


#################
### functions ###
#################

#this will return the user to the menu
pause()
{
    read -p $'\e[38;5;208mPress [Enter] to continue down the foxhole...\e[0m'
}

#this choice will allow the script to update system
updateSystem()
{
    sudo apt update && sudo apt upgrade

    pause 
}

#this choice will allow the script to change hostname
hostnameChange()
{
    local userChoice=$1
    local userHostname=$(hostname)

    echo -e ${YELLOW}"Your existing hostname is:" `hostname`${NT}
    sleep 1
    read -p "Type what you want your hostname to be: " userChoice
    sudo hostnamectl set-hostname $userChoice
    sleep 1
    echo -e ${YELLOW}"Hostname set"
    sleep 1
    echo -e "changing /etc/hosts..."${NT}
    sleep 1
    sudo sed -i "s/$userHostname/$userChoice/g" /etc/hosts
    sleep 1
    echo -e ${YELLOW}"Checking status..."${NT}
    sleep 1.5
    
    echo -e ${ORANGE}"#####################################"${NT}
    echo -e ${YELLOW}"    ${BOLD}---+++--- hostname ---+++---"${NT}
    echo -e ${ORANGE}"#####################################"${NT}
    hostnamectl status | grep host*
    echo -e ${ORANGE}"#####################################"${NT}
    echo -e ${YELLOW}"   ${BOLD}---+++--- /etc/hosts ---+++---"${NT}
    echo -e ${ORANGE}"#####################################"${NT}
    < /etc/hosts grep 127*
    
    pause 
    
}

#this choice will allow the script to change the pi user password
piPasswordChange()
{
    passwd
    pause
}

#this choice will guide the customer to the raspi-config menu to enable SSH
enableSSH()
{
    echo -e ${ORANGE}"####################################################"
    echo -e ${RED}" --+++++-- Configure SSH and Localization --+++++--"
    echo -e ${ORANGE}"####################################################"
    sleep 1
    echo -e ${ORANGE}"When the configuration menu opens go to"${NT}
    echo -e ${RED}"5)Interface Options>P2 SSH and enable it"${NT}
    pause
    sudo raspi-config
    pause
}

#this choice will mount usb storage
mountStorage()
{
    local usbUUID=$1

    echo -e ${YELLOW}"Making directory /media/foxlab..."
    sudo mkdir /media/foxlab
    sleep 0.5
    echo -e ${ORANGE}"done"${NT}
    sleep 0.5
    echo -e ${YELLOW}"checking status..."${NT}
    sleep 0.5
    ls /media
    sleep 0.2
    
    echo -e ${YELLOW}"Changing owner of /media/foxlab..."
    sudo chown pi /media/foxlab
    sleep 0.5
    echo -e ${ORANGE}"done"${NT}
    sleep 0.5

    echo -e ${YELLOW}"Formating the USB storage..."
    sudo sudo mkfs.ext4 /dev/sda1
    sleep 0.5
    echo -e ${ORANGE}"done"${NT}
    sleep 0.5

    echo -e ${YELLOW}"Mounting USB storage to /media/foxlab..."
    sudo mount /dev/sda1 /media/foxlab
    sleep 0.5
    echo -e ${ORANGE}"done"${NT}
    sleep 0.5

    echo -e ${YELLOW}"FoxLab needs your help grabbing the USB storage UUID..."${NT}
    sleep 2
    echo -e                      ${ORANGE}"####################################################"
    echo -e ${BLINK}${REDB}${WHITE}${BOLD} "     copy ONLY what is in quotes after UUID=       "${NT}
    echo -e                      ${ORANGE}"####################################################"${NT}
    sleep 3
    sudo blkid /dev/sda1 | awk '{print $2}'
    echo -e ${ORANGE}"####################################################"${NT}

    sleep 0.5
    read -p $'\e[38;5;11mNow insert the UUID here:\e[0m' usbUUID
    sleep 0.5
    echo -e $usbUUID ${YELLOW}"was inserted"${NT}
    sleep 1
    echo -e "Adding UUID to /etc/fstab..."${NT}
    sleep 0.5
    echo "UUID="$usbUUID"  /media/foxlab  ext4  defaults  0  0" | sudo tee -a /etc/fstab
    sleep 1
    echo -e ${ORANGE}"done"${NT}
    sleep 0.5
    echo -e ${YELLOW}"Checking status..."${NT}
    sleep 0.5

    cat /etc/fstab
    echo -e ${ORANGE}"done"${NT}
    pause
}

#this choice will increase STACK size
stackSize()
{
    echo -e ${YELLOW}"Checking RAM and swapsize..."${NT}
    sleep 1
    echo -e ${ORANGE}"####################################################"${NT}
    echo -e ${YELLOW}"  ----++++---- Free Memory & Swapsize ----++++----  "${NT}
    echo -e ${ORANGE}"####################################################"${NT}
    free --mega | awk '{print (NR==1?"Type":""), $1, $2, $3, (NR==1?"":$4)}' | column -t
    echo -e ${ORANGE}"####################################################"${NT}
    pause
    echo -e ${YELLOW}"Changing swapsize to 2048..."${NT}
    sudo sed -i '16s/.*/CONF_SWAPSIZE=2048/' /etc/dphys-swapfile
    sleep 1
    echo -e ${ORANGE}"done"${NT}
    sleep 1
    echo -e ${YELLOW}"Restarting swap..."${NT}
    sleep 1
    sudo /etc/init.d/dphys-swapfile restart
    sleep 0.5
    echo -e ${ORANGE}"done"${NT}
    sleep 1
    echo -e ${YELLOW}"Checking status..."${NT}
    sleep 1
    echo -e ${ORANGE}"####################################################"${NT}
    echo -e ${YELLOW}"        ----++++---- Swapsize ----++++----          "${NT}
    echo -e ${ORANGE}"####################################################"${NT}
    sed '14,16!d' /etc/dphys-swapfile
    echo -e ${ORANGE}"####################################################"${NT}
    echo -e ${YELLOW}"  ----++++---- Free Memory & Swapsize ----++++----  "${NT}
    echo -e ${ORANGE}"####################################################"${NT}
    free --mega | awk '{print (NR==1?"Type":""), $1, $2, $3, (NR==1?"":$4)}' | column -t
    echo -e ${ORANGE}"####################################################"${NT}

    pause
}

#this choice will download GitLab and install
gitLabInstall() 
{
    local userIP=$1

    echo -e ${ORANGE}"####################################################"${NT}
    echo -e ${YELLOW}"   ----++++---- Downloading GitLab ----++++----     "${NT}
    echo -e ${ORANGE}"####################################################"${NT}
    sleep 1
    curl -Lo gitlab-ce_12.6.2-ce.0_armhf.deb https://packages.gitlab.com/gitlab/raspberry-pi2/packages/raspbian/stretch/gitlab-ce_12.6.2-ce.0_armhf.deb/download.deb

    echo -e ${ORANGE}"####################################################"${NT}
    echo -e ${YELLOW}"   ----++++---- Installing GitLab ----++++----      "${NT}
    echo -e ${ORANGE}"####################################################"${NT}
    sleep 1
    sudo dpkg -i gitlab-ce_12.6.2-ce.0_armhf.deb
    echo -e ${ORANGE}"####################################################"${NT}
    echo -e ${YELLOW}"  ----++++---- GitLab Configuration ----++++----    "${NT}
    echo -e ${ORANGE}"####################################################"${NT}

    echo -e ${YELLOW}"Checking IP address..."
    sleep 1
    echo -e ${ORANGE}"####################################################"${NT}
    echo -e ${YELLOW}"     ----++++---- Raspberry Pi IP ----++++----      "${NT}
    echo -e ${ORANGE}"####################################################"${NT}
    echo "http://" | hostname -I
    echo -e ${ORANGE}"####################################################"${NT}
    echo -e ${BLINK}${REDB}${WHITE}${BOLD}"  COPY the IP address above with http:// before it  "${NT}
    echo -e ${ORANGE}"####################################################"${NT}
    sleep 1

    read -p $'\e[38;5;11mInsert the IP address here:\e[0m' userIP
    sleep 1
    echo -e $userIP ${YELLOW}"was inserted"${NT}
    sleep 1
    echo -e ${YELLOW}"Adding IP address to /etc/gitlab/gitlab.rb..."${NT}
    sleep 1
    sudo sed -i "s,http://gitlab.example.com,$userIP,g" /etc/gitlab/gitlab.rb
    echo -e ${ORANGE}"done"${NT}
    sleep 1
    echo -e ${YELLOW}"Checking status..."${NT}
    sleep 1
    sudo sed '23!d' /etc/gitlab/gitlab.rb
    echo -e ${ORANGE}"done"${NT}
    sleep 1

    echo -e ${YELLOW}"Changing unicorn worker processes to 2..."
    sleep 1
    sudo sed -i "750 s/.*/unicorn['worker_processes'] = 2/" /etc/gitlab/gitlab.rb
    echo -e ${ORANGE}"done"${NT}
    sleep 1
    echo -e ${YELLOW}"Checking status..."${NT}
    sleep 1
    sudo sed '750!d' /etc/gitlab/gitlab.rb
    echo -e ${ORANGE}"done"${NT}
    sleep 1

    echo -e ${YELLOW}"Changing the amount of concurrency in your Sidekiq process..."
    sleep 1
    sudo sed -i "814 s/.*/sidekiq['concurrency'] = 9/" /etc/gitlab/gitlab.rb
    echo -e ${ORANGE}"done"${NT}
    sleep 1
    echo -e ${YELLOW}"Checking status..."${NT}
    sleep 1
    sudo sed '814!d' /etc/gitlab/gitlab.rb
    echo -e ${ORANGE}"done"${NT}
    sleep 1

    echo -e ${YELLOW}"Changing where GitLab data is stored..."
    sleep 1
    sudo sed -i '438 s,#, ,' /etc/gitlab/gitlab.rb
    sudo sed -i '439 s,#, ,' /etc/gitlab/gitlab.rb
    sudo sed -i '440 s,.*,     "path" => "/media/foxlab/git-data",' /etc/gitlab/gitlab.rb
    sudo sed -i '441 s,#, ,' /etc/gitlab/gitlab.rb
    sudo sed -i '442 s,#, ,' /etc/gitlab/gitlab.rb
    echo -e ${ORANGE}"done"${NT}
    sleep 1
    echo -e ${YELLOW}"Checking status..."${NT}
    sleep 1
    sudo sed '438,442!d' /etc/gitlab/gitlab.rb
    echo -e ${ORANGE}"done"${NT}
    sleep 1

    echo -e ${YELLOW}"If everything above looks good then"
    echo -e ${YELLOW}"continue down the fox hole otherwise"
    echo -e ${YELLOW}"press [CTL+C] to end the program"
    pause
    echo -e ${YELLOW}"Reconfiguring /etc/gitlab/gitlab.rb..."
    sudo gitlab-ctl reconfigure
    echo -e ${ORANGE}"done"${NT}
    pause
}

#this choice will check if gitlab is running
gitLabCheck()
{
    sudo gitlab-ctl status 

    pause
}


#this will display the choices in a menu format
foxLabMenu()
{
    clear                     

    #FoxLabPi ASCII Art
    echo -e ${ORANGE}"               *.                  *."
    echo -e ${ORANGE}"              ***                 ***"
    echo -e ${ORANGE}"             *****               *****"
    echo -e ${ORANGE}"            .******             *******"
    echo -e ${ORANGE}"            ********            ********"
    echo -e ${ORANGE}"           ,,,,,,,,,***********,,,,,,,,,"
    echo -e ${ORANGE}"          ,,,,,,,,,,,*********,,,,,,,,,,,"
    echo -e ${ORANGE}"          .,,,,,,,,,,,*******,,,,,,,,,,,,"
    echo -e ${ORANGE}"              ,,,,,,,,,*****,,,,,,,,,."
    echo -e ${ORANGE}"                 ,,,,,,,****,,,,,,"
    echo -e ${ORANGE}"                    .,,,***,,,,"
    echo -e ${ORANGE}"                        ,*,."

    echo -e ${ORANGE}"       ___           __       _         ${RED}___ _"
    echo -e ${ORANGE}"      / __\____  __ / /  __ _| |__     ${RED}/ _ (_)"
    echo -e ${ORANGE}"     / _\/ _ \ \/ // /  / _  |  _ \   ${RED}/ /_)/ |"
    echo -e ${ORANGE}"    / / | (_) >  </ /__| (_| | |_) | ${RED}/ ___/| |"
    echo -e ${ORANGE}"    \/   \___/_/\_\____/\__,_|_.__/  ${RED}\/    |_|"${NT}

    echo -e ${PURPLE}"####################################################"
    echo -e "  ${WHITE}----${PURPLE}+++++++${WHITE}---- ${BOLD}${ORANGE}FoxLab ${RED}Pi ${YELLOW}Menu ${NT}${WHITE}----${PURPLE}+++++++${WHITE}----    "
    echo -e ${PURPLE}"####################################################"${NT}
    echo -e ${PURPLE}"1${ORANGE}) ${LITEGREY}Update System ${LITEPURPLE}- - - - - - - - - - -${ORANGE}(${ORANGE}Recommended${ORANGE})"
    echo -e ${PURPLE}"2${ORANGE}) ${LITEGREY}Change Pi Password ${LITEPURPLE}- - - - - - - - ${ORANGE}(${ORANGE}Recommended${ORANGE})"
    echo -e ${PURPLE}"3${ORANGE}) ${LITEGREY}Change Hostname ${LITEPURPLE}- - - - - - - - - -${ORANGE}(${ORANGE}Recommended${ORANGE})"
    echo -e ${PURPLE}"4${ORANGE}) ${LITEGREY}Enable SSH ${LITEPURPLE}- - - - - - - - - - - - ${ORANGE}(${YELLOW}Optional${ORANGE})"
    echo -e ${PURPLE}"5${ORANGE}) ${LITEGREY}Mount USB Storage ${LITEPURPLE}- - - - - - - - -${ORANGE}(${ORANGE}Recommended${ORANGE})"
    echo -e ${PURPLE}"6${ORANGE}) ${LITEGREY}Increase STACK size ${LITEPURPLE}- - - - - - - -${ORANGE}(${ORANGE}Recommended${ORANGE})"
    echo -e ${PURPLE}"7${ORANGE}) ${LITEGREY}Download & install GitLab ${LITEPURPLE}- - - - -${ORANGE}(${RED}Necessary${ORANGE})"
    echo -e ${PURPLE}"8${ORANGE}) ${LITEGREY}Check that GitLab is running ${LITEPURPLE}- - - ${ORANGE}(${YELLOW}Optional${ORANGE})"
    echo -e ${PURPLE}"9${ORANGE}) ${LITEGREY}Exit"
}

#this will read the users input
readOptions()
{
    local userChoice
    read -p $'\e[38;5;208mWhich foxhole do you want to tumble down...\e[0m' userChoice
    case $userChoice in
        1) updateSystem ;;
        2) piPasswordChange ;;
        3) hostnameChange ;;
        4) enableSSH ;;
        5) mountStorage ;;
        6) stackSize ;;
        7) gitLabInstall ;;
        8) gitLabCheck ;;
        9) exit 0 ;;
        *) echo -e ${BLINK}${REDB}${WHITE}"WARNING${NT}${ORANGE}...You have traveled down the wrong foxhole"${NT} && sleep 4 ;;
    esac
}

while true
do
    foxLabMenu
    readOptions
done
